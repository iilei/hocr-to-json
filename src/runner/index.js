import fs from 'fs'
import JSONstringify from 'json-stable-stringify'
import { transform } from 'camaro'
import {
  castToArrayOfNumbers,
  castToMatrix,
  constants,
  dropProp,
  parseFloatOrNull,
  propConversionMerge,
  traverse,
  traverseFactory,
  unQuote,
} from '../modules'
import { bestMatch } from '../modules/view'

import * as views from '../views'

const xml = fs.readFileSync('stub/phototest.hocr', 'utf-8')

const { numberProps, seperator } = constants

const runner = async () => {
  // 2-Step process; first metadata, then decide which _meta to apply
  const meta = await transform(xml, views._meta)

  const [engine, version] = meta.ocrSystem.split(seperator)

  const view = bestMatch(views, engine, version)

  const body = await transform(xml, view)

  const result = { ...meta, ...body }
  traverseFactory(result, propConversionMerge, 'title')
  traverseFactory(result, dropProp, 'title')
  traverse(result, castToMatrix, 'bbox')
  traverse(result, castToArrayOfNumbers, 'baseline')
  traverse(result, unQuote, 'image')
  traverse(result, parseFloatOrNull, ...numberProps)

  fs.writeFileSync('stub/phototest.json', JSONstringify(result, { space: '  ' }), {
    encoding: 'utf-8',
  })

  return result
}

export default runner
