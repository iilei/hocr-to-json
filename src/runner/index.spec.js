import runner from './index'

const defautArgs = ['hocr-json', '*.hocr']

describe('runner', () => {
  it('should invoke the desired task', async () => {
    const result = await runner(...defautArgs)
    expect(result).toMatchSnapshot()

    // const taskMock = require('../tasks').myTask
    // const { appendBasePath, setAuthToken } = require('../modules/axiosInstance')
    // const runner = require('./index').default
    //
    // await runner('myTask', '/foo')
    //
    // expect(setAuthToken.mock.calls[0][0]).toBe('JWT')
    // expect(appendBasePath.mock.calls[0][0]).toBe('/projects/xyz/')
    // expect(taskMock.mock.calls[0][0]).toEqual(
    //   expect.objectContaining({
    //     'client-id': 'test',
    //     'pull-to': '/abs-path/foo/intl/pull/<locale>.<hash:6>.json',
    //   }),
    // )
  })
})
