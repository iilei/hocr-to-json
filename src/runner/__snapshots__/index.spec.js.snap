// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runner should invoke the desired task with defaults 1`] = `
"{
  \\"contentType\\": \\"text/html;charset=utf-8\\",
  \\"ocrCapabilities\\": \\"ocr_page ocr_carea ocr_par ocr_line ocrx_word ocrp_wconf\\",
  \\"ocrSystem\\": \\"tesseract 4.1.0-rc1-752-g8b69\\",
  \\"pages\\": [
    {
      \\"bbox\\": [
        [
          0,
          0
        ],
        [
          640,
          480
        ]
      ],
      \\"careas\\": [
        {
          \\"bbox\\": [
            [
              36,
              92
            ],
            [
              618,
              361
            ]
          ],
          \\"id\\": \\"block_1_1\\",
          \\"pars\\": [
            {
              \\"id\\": \\"par_1_1\\",
              \\"lang\\": \\"eng\\",
              \\"lines\\": [
                {
                  \\"baseline\\": [
                    0,
                    -6
                  ],
                  \\"bbox\\": [
                    [
                      36,
                      92
                    ],
                    [
                      580,
                      122
                    ]
                  ],
                  \\"id\\": \\"line_1_1\\",
                  \\"words\\": [
                    {
                      \\"bbox\\": [
                        [
                          36,
                          92
                        ],
                        [
                          96,
                          116
                        ]
                      ],
                      \\"content\\": \\"This\\",
                      \\"id\\": \\"word_1_1\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          109,
                          92
                        ],
                        [
                          129,
                          116
                        ]
                      ],
                      \\"content\\": \\"is\\",
                      \\"id\\": \\"word_1_2\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          141,
                          98
                        ],
                        [
                          156,
                          116
                        ]
                      ],
                      \\"content\\": \\"a\\",
                      \\"id\\": \\"word_1_3\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          169,
                          92
                        ],
                        [
                          201,
                          116
                        ]
                      ],
                      \\"content\\": \\"lot\\",
                      \\"id\\": \\"word_1_4\\",
                      \\"xWconf\\": 90
                    },
                    {
                      \\"bbox\\": [
                        [
                          212,
                          92
                        ],
                        [
                          240,
                          116
                        ]
                      ],
                      \\"content\\": \\"of\\",
                      \\"id\\": \\"word_1_5\\",
                      \\"xWconf\\": 93
                    },
                    {
                      \\"bbox\\": [
                        [
                          251,
                          92
                        ],
                        [
                          282,
                          116
                        ]
                      ],
                      \\"content\\": \\"12\\",
                      \\"id\\": \\"word_1_6\\",
                      \\"xWconf\\": 93
                    },
                    {
                      \\"bbox\\": [
                        [
                          296,
                          92
                        ],
                        [
                          364,
                          122
                        ]
                      ],
                      \\"content\\": \\"point\\",
                      \\"id\\": \\"word_1_7\\",
                      \\"xWconf\\": 90
                    },
                    {
                      \\"bbox\\": [
                        [
                          374,
                          93
                        ],
                        [
                          427,
                          116
                        ]
                      ],
                      \\"content\\": \\"text\\",
                      \\"id\\": \\"word_1_8\\",
                      \\"xWconf\\": 93
                    },
                    {
                      \\"bbox\\": [
                        [
                          437,
                          93
                        ],
                        [
                          463,
                          116
                        ]
                      ],
                      \\"content\\": \\"to\\",
                      \\"id\\": \\"word_1_9\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          474,
                          93
                        ],
                        [
                          526,
                          116
                        ]
                      ],
                      \\"content\\": \\"test\\",
                      \\"id\\": \\"word_1_10\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          536,
                          92
                        ],
                        [
                          580,
                          116
                        ]
                      ],
                      \\"content\\": \\"the\\",
                      \\"id\\": \\"word_1_11\\",
                      \\"xWconf\\": 92
                    }
                  ],
                  \\"xAscenders\\": 6,
                  \\"xDescenders\\": 6,
                  \\"xSize\\": 30
                },
                {
                  \\"baseline\\": [
                    0,
                    -7
                  ],
                  \\"bbox\\": [
                    [
                      36,
                      126
                    ],
                    [
                      618,
                      157
                    ]
                  ],
                  \\"id\\": \\"line_1_2\\",
                  \\"words\\": [
                    {
                      \\"bbox\\": [
                        [
                          36,
                          132
                        ],
                        [
                          81,
                          150
                        ]
                      ],
                      \\"content\\": \\"ocr\\",
                      \\"id\\": \\"word_1_12\\",
                      \\"xWconf\\": 90
                    },
                    {
                      \\"bbox\\": [
                        [
                          91,
                          126
                        ],
                        [
                          160,
                          150
                        ]
                      ],
                      \\"content\\": \\"code\\",
                      \\"id\\": \\"word_1_13\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          172,
                          126
                        ],
                        [
                          223,
                          150
                        ]
                      ],
                      \\"content\\": \\"and\\",
                      \\"id\\": \\"word_1_14\\",
                      \\"xWconf\\": 90
                    },
                    {
                      \\"bbox\\": [
                        [
                          236,
                          132
                        ],
                        [
                          286,
                          150
                        ]
                      ],
                      \\"content\\": \\"see\\",
                      \\"id\\": \\"word_1_15\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          299,
                          126
                        ],
                        [
                          314,
                          150
                        ]
                      ],
                      \\"content\\": \\"if\\",
                      \\"id\\": \\"word_1_16\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          325,
                          126
                        ],
                        [
                          339,
                          150
                        ]
                      ],
                      \\"content\\": \\"it\\",
                      \\"id\\": \\"word_1_17\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          348,
                          126
                        ],
                        [
                          433,
                          150
                        ]
                      ],
                      \\"content\\": \\"works\\",
                      \\"id\\": \\"word_1_18\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          445,
                          132
                        ],
                        [
                          478,
                          150
                        ]
                      ],
                      \\"content\\": \\"on\\",
                      \\"id\\": \\"word_1_19\\",
                      \\"xWconf\\": 87
                    },
                    {
                      \\"bbox\\": [
                        [
                          500,
                          126
                        ],
                        [
                          529,
                          150
                        ]
                      ],
                      \\"content\\": \\"all\\",
                      \\"id\\": \\"word_1_20\\",
                      \\"xWconf\\": 87
                    },
                    {
                      \\"bbox\\": [
                        [
                          541,
                          127
                        ],
                        [
                          618,
                          157
                        ]
                      ],
                      \\"content\\": \\"types\\",
                      \\"id\\": \\"word_1_21\\",
                      \\"xWconf\\": 92
                    }
                  ],
                  \\"xAscenders\\": 6,
                  \\"xDescenders\\": 7,
                  \\"xSize\\": 31
                },
                {
                  \\"baseline\\": [
                    0,
                    0
                  ],
                  \\"bbox\\": [
                    [
                      36,
                      160
                    ],
                    [
                      223,
                      184
                    ]
                  ],
                  \\"id\\": \\"line_1_3\\",
                  \\"words\\": [
                    {
                      \\"bbox\\": [
                        [
                          36,
                          160
                        ],
                        [
                          64,
                          184
                        ]
                      ],
                      \\"content\\": \\"of\\",
                      \\"id\\": \\"word_1_22\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          72,
                          160
                        ],
                        [
                          113,
                          184
                        ]
                      ],
                      \\"content\\": \\"file\\",
                      \\"id\\": \\"word_1_23\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          123,
                          160
                        ],
                        [
                          223,
                          184
                        ]
                      ],
                      \\"content\\": \\"format.\\",
                      \\"id\\": \\"word_1_24\\",
                      \\"xWconf\\": 91
                    }
                  ],
                  \\"xAscenders\\": 6,
                  \\"xDescenders\\": 7.2148418,
                  \\"xSize\\": 31.214842
                }
              ]
            },
            {
              \\"id\\": \\"par_1_2\\",
              \\"lang\\": \\"eng\\",
              \\"lines\\": [
                {
                  \\"baseline\\": [
                    0,
                    -7
                  ],
                  \\"bbox\\": [
                    [
                      36,
                      194
                    ],
                    [
                      585,
                      225
                    ]
                  ],
                  \\"id\\": \\"line_1_4\\",
                  \\"words\\": [
                    {
                      \\"bbox\\": [
                        [
                          36,
                          194
                        ],
                        [
                          91,
                          218
                        ]
                      ],
                      \\"content\\": \\"The\\",
                      \\"id\\": \\"word_1_25\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          102,
                          194
                        ],
                        [
                          177,
                          224
                        ]
                      ],
                      \\"content\\": \\"quick\\",
                      \\"id\\": \\"word_1_26\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          189,
                          194
                        ],
                        [
                          274,
                          218
                        ]
                      ],
                      \\"content\\": \\"brown\\",
                      \\"id\\": \\"word_1_27\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          287,
                          194
                        ],
                        [
                          339,
                          225
                        ]
                      ],
                      \\"content\\": \\"dog\\",
                      \\"id\\": \\"word_1_28\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          348,
                          194
                        ],
                        [
                          456,
                          225
                        ]
                      ],
                      \\"content\\": \\"jumped\\",
                      \\"id\\": \\"word_1_29\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          468,
                          200
                        ],
                        [
                          531,
                          218
                        ]
                      ],
                      \\"content\\": \\"over\\",
                      \\"id\\": \\"word_1_30\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          540,
                          194
                        ],
                        [
                          585,
                          218
                        ]
                      ],
                      \\"content\\": \\"the\\",
                      \\"id\\": \\"word_1_31\\",
                      \\"xWconf\\": 92
                    }
                  ],
                  \\"xAscenders\\": 6,
                  \\"xDescenders\\": 7,
                  \\"xSize\\": 31
                },
                {
                  \\"baseline\\": [
                    0,
                    -7
                  ],
                  \\"bbox\\": [
                    [
                      37,
                      228
                    ],
                    [
                      585,
                      259
                    ]
                  ],
                  \\"id\\": \\"line_1_5\\",
                  \\"words\\": [
                    {
                      \\"bbox\\": [
                        [
                          37,
                          228
                        ],
                        [
                          92,
                          259
                        ]
                      ],
                      \\"content\\": \\"lazy\\",
                      \\"id\\": \\"word_1_32\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          103,
                          228
                        ],
                        [
                          153,
                          252
                        ]
                      ],
                      \\"content\\": \\"fox.\\",
                      \\"id\\": \\"word_1_33\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          165,
                          228
                        ],
                        [
                          220,
                          252
                        ]
                      ],
                      \\"content\\": \\"The\\",
                      \\"id\\": \\"word_1_34\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          232,
                          228
                        ],
                        [
                          307,
                          258
                        ]
                      ],
                      \\"content\\": \\"quick\\",
                      \\"id\\": \\"word_1_35\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          319,
                          228
                        ],
                        [
                          404,
                          252
                        ]
                      ],
                      \\"content\\": \\"brown\\",
                      \\"id\\": \\"word_1_36\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          417,
                          228
                        ],
                        [
                          468,
                          259
                        ]
                      ],
                      \\"content\\": \\"dog\\",
                      \\"id\\": \\"word_1_37\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          478,
                          228
                        ],
                        [
                          585,
                          259
                        ]
                      ],
                      \\"content\\": \\"jumped\\",
                      \\"id\\": \\"word_1_38\\",
                      \\"xWconf\\": 92
                    }
                  ],
                  \\"xAscenders\\": 6,
                  \\"xDescenders\\": 7,
                  \\"xSize\\": 31
                },
                {
                  \\"baseline\\": [
                    0,
                    -7
                  ],
                  \\"bbox\\": [
                    [
                      36,
                      262
                    ],
                    [
                      597,
                      293
                    ]
                  ],
                  \\"id\\": \\"line_1_6\\",
                  \\"words\\": [
                    {
                      \\"bbox\\": [
                        [
                          36,
                          268
                        ],
                        [
                          99,
                          286
                        ]
                      ],
                      \\"content\\": \\"over\\",
                      \\"id\\": \\"word_1_39\\",
                      \\"xWconf\\": 90
                    },
                    {
                      \\"bbox\\": [
                        [
                          109,
                          262
                        ],
                        [
                          153,
                          286
                        ]
                      ],
                      \\"content\\": \\"the\\",
                      \\"id\\": \\"word_1_40\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          165,
                          262
                        ],
                        [
                          221,
                          293
                        ]
                      ],
                      \\"content\\": \\"lazy\\",
                      \\"id\\": \\"word_1_41\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          231,
                          262
                        ],
                        [
                          281,
                          286
                        ]
                      ],
                      \\"content\\": \\"fox.\\",
                      \\"id\\": \\"word_1_42\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          294,
                          262
                        ],
                        [
                          349,
                          286
                        ]
                      ],
                      \\"content\\": \\"The\\",
                      \\"id\\": \\"word_1_43\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          360,
                          262
                        ],
                        [
                          435,
                          292
                        ]
                      ],
                      \\"content\\": \\"quick\\",
                      \\"id\\": \\"word_1_44\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          447,
                          262
                        ],
                        [
                          532,
                          286
                        ]
                      ],
                      \\"content\\": \\"brown\\",
                      \\"id\\": \\"word_1_45\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          545,
                          262
                        ],
                        [
                          597,
                          293
                        ]
                      ],
                      \\"content\\": \\"dog\\",
                      \\"id\\": \\"word_1_46\\",
                      \\"xWconf\\": 90
                    }
                  ],
                  \\"xAscenders\\": 6,
                  \\"xDescenders\\": 7,
                  \\"xSize\\": 31
                },
                {
                  \\"baseline\\": [
                    0,
                    -7
                  ],
                  \\"bbox\\": [
                    [
                      43,
                      296
                    ],
                    [
                      561,
                      327
                    ]
                  ],
                  \\"id\\": \\"line_1_7\\",
                  \\"words\\": [
                    {
                      \\"bbox\\": [
                        [
                          43,
                          296
                        ],
                        [
                          150,
                          327
                        ]
                      ],
                      \\"content\\": \\"jumped\\",
                      \\"id\\": \\"word_1_47\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          162,
                          302
                        ],
                        [
                          226,
                          320
                        ]
                      ],
                      \\"content\\": \\"over\\",
                      \\"id\\": \\"word_1_48\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          235,
                          296
                        ],
                        [
                          279,
                          320
                        ]
                      ],
                      \\"content\\": \\"the\\",
                      \\"id\\": \\"word_1_49\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          292,
                          296
                        ],
                        [
                          347,
                          327
                        ]
                      ],
                      \\"content\\": \\"lazy\\",
                      \\"id\\": \\"word_1_50\\",
                      \\"xWconf\\": 88
                    },
                    {
                      \\"bbox\\": [
                        [
                          357,
                          296
                        ],
                        [
                          407,
                          320
                        ]
                      ],
                      \\"content\\": \\"fox.\\",
                      \\"id\\": \\"word_1_51\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          420,
                          296
                        ],
                        [
                          475,
                          320
                        ]
                      ],
                      \\"content\\": \\"The\\",
                      \\"id\\": \\"word_1_52\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          486,
                          296
                        ],
                        [
                          561,
                          326
                        ]
                      ],
                      \\"content\\": \\"quick\\",
                      \\"id\\": \\"word_1_53\\",
                      \\"xWconf\\": 91
                    }
                  ],
                  \\"xAscenders\\": 6,
                  \\"xDescenders\\": 7,
                  \\"xSize\\": 31
                },
                {
                  \\"baseline\\": [
                    0,
                    -7
                  ],
                  \\"bbox\\": [
                    [
                      37,
                      330
                    ],
                    [
                      561,
                      361
                    ]
                  ],
                  \\"id\\": \\"line_1_8\\",
                  \\"words\\": [
                    {
                      \\"bbox\\": [
                        [
                          37,
                          330
                        ],
                        [
                          122,
                          354
                        ]
                      ],
                      \\"content\\": \\"brown\\",
                      \\"id\\": \\"word_1_54\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          135,
                          330
                        ],
                        [
                          187,
                          361
                        ]
                      ],
                      \\"content\\": \\"dog\\",
                      \\"id\\": \\"word_1_55\\",
                      \\"xWconf\\": 92
                    },
                    {
                      \\"bbox\\": [
                        [
                          196,
                          330
                        ],
                        [
                          304,
                          361
                        ]
                      ],
                      \\"content\\": \\"jumped\\",
                      \\"id\\": \\"word_1_56\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          316,
                          336
                        ],
                        [
                          379,
                          354
                        ]
                      ],
                      \\"content\\": \\"over\\",
                      \\"id\\": \\"word_1_57\\",
                      \\"xWconf\\": 90
                    },
                    {
                      \\"bbox\\": [
                        [
                          388,
                          330
                        ],
                        [
                          433,
                          354
                        ]
                      ],
                      \\"content\\": \\"the\\",
                      \\"id\\": \\"word_1_58\\",
                      \\"xWconf\\": 91
                    },
                    {
                      \\"bbox\\": [
                        [
                          445,
                          330
                        ],
                        [
                          500,
                          361
                        ]
                      ],
                      \\"content\\": \\"lazy\\",
                      \\"id\\": \\"word_1_59\\",
                      \\"xWconf\\": 89
                    },
                    {
                      \\"bbox\\": [
                        [
                          511,
                          330
                        ],
                        [
                          561,
                          354
                        ]
                      ],
                      \\"content\\": \\"fox.\\",
                      \\"id\\": \\"word_1_60\\",
                      \\"xWconf\\": 90
                    }
                  ],
                  \\"xAscenders\\": 6,
                  \\"xDescenders\\": 7,
                  \\"xSize\\": 31
                }
              ]
            }
          ]
        }
      ],
      \\"id\\": \\"page_1\\",
      \\"image\\": \\"phototest.tif\\",
      \\"ppageno\\": 0
    }
  ]
}"
`;
